# Linear Regression with _statsmodels_

We have seen that sometimes different data sets appear to vary together.  Statistical methods such as regression attempt answer that question.

__Linear Regression__ attempts to model this relationship between a __response__ or __dependent variable__ and one or more __independent variables__.

Linear regression may be used to a) quantify the strength of the relationship between data series and b) forecast future values of the dependent variable based on the independent data.

__NOTE__: Regression analysis is a deep and complex subject: this is hopefully a good practical introduction to the topic but there are many subtleties and I encourage you to do more in-depth coursework to build on this foundation.

### Definition
A full treatment of regression is well beyond the time we have, but to put it simply, a linear regression model assumes that there is some relationship between the dependent variable series $\vec{y}$ and some matrix $\boldsymbol{X}$ of independent series where each series is related to $\vec{y}$ by a matrix of coefficients $\boldsymbol{\beta}$:

### $$ \vec{y} = \beta\boldsymbol{X} $$

For a _simple regression_ with one independent variable, this looks exactly like a linear function:

### $$ \hat{y} = \beta\vec{x} + c  $$

where $\hat{y}$ is the estimate of $\vec{y}$ and $c$ is the y-intercept of the function.

The goal of the regression analysis is to find the $\beta$ coefficient that best fits the data. One method that we'll look at today, __ordinary least squares__, does this by minimizing the distance between the estimated $\hat{y}$ and the true values $\vec{y}$.

Let's look at an example now.

### Ordinary Least Squares Regression with statsmodels: an example

<sub>(_adapted from [statsmodels.org](https://www.statsmodels.org/devel/)_)</sub>

In [1]:
import statsmodels.api as sm
import statsmodels.formula.api as smf
import numpy as np

# let's get our copper data from last week
copper = sm.datasets.copper.load_pandas().data
years = np.arange(1951, 1976)
copper.index = years

Let's look at this data again, and think about what kinds of economic relationships.

In [2]:
copper

WORLDCONSUMPTION  COPPERPRICE  INCOMEINDEX  ALUMPRICE  INVENTORYINDEX  \
1951            3173.0        26.56         0.70      19.76            0.98   
1952            3281.1        27.31         0.71      20.78            1.04   
1953            3135.7        32.95         0.72      22.55            1.05   
1954            3359.1        33.90         0.70      23.06            0.97   
1955            3755.1        42.70         0.74      24.93            1.02   
1956            3875.9        46.11         0.74      26.50            1.04   
1957            3905.7        31.70         0.74      27.24            0.98   
1958            3957.6        27.23         0.72      26.21            0.98   
1959            4279.1        32.89         0.75      26.09            1.03   
1960            4627.9        33.78         0.77      27.40            1.03   
1961            4910.2        31.66         0.76      26.94            0.98   
1962            4908.4        32.28         0.79      25.18            1.00   
1963            5327.9        32.38         0.83      23.94            0.97   
1964            5878.4        33.75         0.85      25.07            1.03   
1965            6075.2        36.25         0.89      25.37            1.08   
1966            6312.7        36.24         0.93      24.55            1.05   
1967            6056.8        38.23         0.95      24.98            1.03   
1968            6375.9        40.83         0.99      24.96            1.03   
1969            6974.3        44.62         1.00      25.52            0.99   
1970            7101.6        52.27         1.00      26.01            1.00   
1971            7071.7        45.16         1.02      25.46            0.96   
1972            7754.8        42.50         1.07      22.17            0.97   
1973            8480.3        43.70         1.12      18.56            0.98   
1974            8105.2        47.88         1.10      21.32            1.01   
1975            7157.2        36.33         1.07      22.75            0.94   

      TIME  
1951   1.0  
1952   2.0  
1953   3.0  
1954   4.0  
1955   5.0  
1956   6.0  
1957   7.0  
1958   8.0  
1959   9.0  
1960  10.0  
1961  11.0  
1962  12.0  
1963  13.0  
1964  14.0  
1965  15.0  
1966  16.0  
1967  17.0  
1968  18.0  
1969  19.0  
1970  20.0  
1971  21.0  
1972  22.0  
1973  23.0  
1974  24.0  
1975  25.0

We first need to define the variables that we want to use in our $\hat{y} = \beta\vec{x}$.  In statsmodels for OLS, we do this by specifying the model as:

<div style="text-align:center"> $\hat{y}$ ~ $\vec{x}$

In [3]:
# R model formulation, provides intercept by default
results = smf.ols('WORLDCONSUMPTION ~ INCOMEINDEX', data=copper).fit()

In [4]:
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:       WORLDCONSUMPTION   R-squared:                       0.956
Model:                            OLS   Adj. R-squared:                  0.955
Method:                 Least Squares   F-statistic:                     504.6
Date:                Sun, 31 May 2020   Prob (F-statistic):           3.79e-17
Time:                        21:30:04   Log-Likelihood:                -181.31
No. Observations:                  25   AIC:                             366.6
Df Residuals:                      23   BIC:                             369.1
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept   -4365.8514    442.034     -9.877      

In [ ]:
print(results.params)

In [ ]:
results.params['INCOMEINDEX']

### So what does all this stuff mean?

Let's look at the output equation itself, including the intercept and coefficient. Our output function looks like

$$
\mbox{WORLDCONSUMPTION} = (1.131\times10^4 * \mbox{INCOMEINDEX}) - 4365.8514
$$

In [ ]:
# we could plot that...
import matplotlib.pyplot as plt

intercept = results.params.Intercept
income_coef = results.params.INCOMEINDEX

rsq = results.rsquared

#plt.plot(income_coef*copper.INCOMEINDEX + intercept, "r", label="Predicted")
plt.plot(results.fittedvalues, "r", label="Predicted")

plt.plot(copper.WORLDCONSUMPTION, "g", label="Actual")

plt.legend()
plt.title("R-squared = {:.3f}".format(rsq))
plt.show()

### Goodness of Fit, Part I

__R-squared__ This describes the "goodness of fit", or well the model explains the variation of the dependent variable. It can be thought of as:

$$ \mbox{R-squared} = \frac{\mbox{explained variation}}{\mbox{total variation}} $$

__AIC__ Aikake Information Criterion - a measure of the distance between fitted values and the unknown "true" model. Lower is better.

__BIC__ Bayesian Information Criterion - similar to AIC. Lower is better.

__Coefficient p-values__ a measure of whether a given independent variable is signficant. Lower is better (generally p < 0.05 is significant)

### Regressing without an intercept

The smf OLS package automatically adds an intercept term to the regression. Sometimes we don't want that. To this we can specify a "-1" in the regression model:

In [ ]:
# R model formulation, provides intercept by default
results2 = smf.ols('WORLDCONSUMPTION ~ INCOMEINDEX - 1', data=copper).fit()

In [ ]:
print(results2.summary())

In [ ]:
print(results2.params)

In [ ]:
# we could plot that...
import matplotlib.pyplot as plt

rsq2 = results2.rsquared
income_coef2 = results2.params.INCOMEINDEX

#plt.plot(income_coef2*copper.INCOMEINDEX, "r", label="Predicted")
plt.plot(results2.fittedvalues, "r", label="Predicted")

plt.plot(copper.WORLDCONSUMPTION, "g", label="Actual")
plt.legend()
plt.title("R-squared = {:.3f}".format(rsq2))
plt.show()

Another way to visualize the goodness of fit is through the use of regression plot.

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(16,5))

# with intercept
ax1.scatter(copper.INCOMEINDEX, copper.WORLDCONSUMPTION)
ax1.plot(copper.INCOMEINDEX,income_coef*copper.INCOMEINDEX + intercept, "r")
ax1.title.set_text("With intercept")

# without intercept
ax2.scatter(copper.INCOMEINDEX, copper.WORLDCONSUMPTION)
ax2.plot(copper.INCOMEINDEX,income_coef2*copper.INCOMEINDEX, "r")
ax2.title.set_text("Without intercept")

plt.show()

## Multiple Regression

We can do the exact same approach but use multiple independent variables in the model

In [ ]:
# R model formulation, provides intercept by default
results_mult = smf.ols('WORLDCONSUMPTION ~ INCOMEINDEX + COPPERPRICE', data=copper).fit()

In [ ]:
print(results_mult.summary())

In [ ]:
params_mult = results_mult.params
print(params_mult)

In [ ]:
# let's plot our actual vs fitted again...
import matplotlib.pyplot as plt

rsq_mult = results_mult.rsquared
income_coef_mult = results_mult.params.INCOMEINDEX

plt.plot(results_mult.fittedvalues, "r", label="Predicted")

plt.plot(copper.WORLDCONSUMPTION, "g", label="Actual")
plt.legend()
plt.title("R-squared = {:.3f}".format(rsq_mult))
plt.show()

### Goodness of Fit, Part II: which one of the above is better given the plots, R-squared, AIC and BIC?

if we examine all the above, which one do we think is a better fit?

In [ ]:
import pandas as pd

# make a dataframe for easy comparison
res1 = {"rsq" : results.rsquared, "rsq-adj" : results.rsquared_adj, "AIC" : results.aic, "BIC" : results.bic }
res2 = {"rsq" : results2.rsquared, "rsq-adj" : results2.rsquared_adj, "AIC" : results2.aic, "BIC" : results2.bic }
res3 = {"rsq" : results_mult.rsquared, "rsq-adj" : results_mult.rsquared_adj, "AIC" : results_mult.aic, "BIC" : results_mult.bic }

all_results = pd.DataFrame([res1, res2, res3], index=["intercept", "no intercept", "multiple"])

In [ ]:
all_results